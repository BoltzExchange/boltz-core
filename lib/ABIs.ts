import ERC20ABI from '../out/ERC20.sol/ERC20.json';
import ERC20SwapABI from '../out/ERC20Swap.sol/ERC20Swap.json';
import EtherSwapABI from '../out/EtherSwap.sol/EtherSwap.json';
import RouterABI from '../out/Router.sol/Router.json';

export const ContractABIs = {
  ERC20: ERC20ABI.abi,
  ERC20Swap: ERC20SwapABI.abi,
  EtherSwap: EtherSwapABI.abi,
  Router: RouterABI.abi,
} as const;
