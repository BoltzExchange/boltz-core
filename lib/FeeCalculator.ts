import Errors from './consts/Errors';
import { OutputType } from './consts/Enums';

export type Input = {
  type: OutputType,

  // In case the input is a Swap this fields have to be set
  swapDetails?: {
    redeemScript: Buffer;
    preimage?: Buffer;
  };
};

export type Output = {
  type: OutputType;

  isSh?: boolean;
};

// Estimations for the vbytes of different PKH inputs
const inputVbytesEstimations = {
  [OutputType.Bech32]: 108 + (41 * 4),
  [OutputType.Compatibility]: 108 + (64 * 4),
  [OutputType.Legacy]: 148 * 4,
};

// Estimations for the vbytes of different PKH and SH outputs
const outputVbytesEstimations = {
  pkh: {
    [OutputType.Bech32]: 31 * 4,
    [OutputType.Legacy]: 34 * 4,
    [OutputType.Taproot]: 43 * 4,
  },
  sh: {
    [OutputType.Bech32]: 44 * 4,
    [OutputType.Legacy]: 32 * 4,
  },
};

const estimateInput = (input: Input) => {
  if (input.type === OutputType.Taproot) {
    throw Errors.TAPROOT_NOT_SUPPORTED;
  }

  if (input.swapDetails) {
    const swapSize = [
      // PUSHDATA opcode
      1,
      // ECDSA signature
      72,
      // PUSHDATA opcode if there is a preimage
      input.swapDetails.preimage ? 1 : 0,
      // preimage if there is one
      input.swapDetails.preimage ? input.swapDetails.preimage.length : 0,
      // Sequence
      4,
      // Redeemscript
      input.swapDetails.redeemScript.length,
    ].reduce((swapSize, n) => swapSize + n);

    switch (input.type) {
      case OutputType.Bech32:
        return 41 * 4 + swapSize;

      case OutputType.Compatibility:
        return ((41 + 35) * 4) + swapSize;

      case OutputType.Legacy:
        return (swapSize + 41) * 4;
    }

    return 0;
  } else {
    return inputVbytesEstimations[input.type];
  }
};

const estimateOutput = (output: Output): number => {
  if (output.type === OutputType.Compatibility) {
    return outputVbytesEstimations.sh[OutputType.Legacy];
  }

  if (output.isSh) {
    return outputVbytesEstimations.sh[output.type];
  } else {
    return outputVbytesEstimations.pkh[output.type];
  }
};

/**
 * Estimates the vbytes of a transaction
 */
export const estimateSize = (inputs: Input[], outputs: Output[]): number => {
  // A raw transaction has always 4 bytes for the version and 4 for the locktime
  let sum = 8 * 4;
  let hasWitness = false;

  inputs.forEach((input) => {
    if (input.type !== OutputType.Legacy) {
      hasWitness = true;
    }

    sum += estimateInput(input);
  });

  outputs.forEach((output) => {
    sum += estimateOutput(output);
  });

  if (hasWitness) {
    sum += 8;
  }

  // Divide the weight by 4 and round it up to the next integer
  return Math.ceil(sum / 4);
};

/**
 * Estimates the amount of satoshis that should be paid as fee
 */
export const estimateFee = (satsPerVbyte: number, inputs: Input[], outputs: Output[]): number => {
  const size = estimateSize(inputs, outputs);

  return size * satsPerVbyte;
};
