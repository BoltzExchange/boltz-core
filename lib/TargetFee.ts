import { Transaction } from 'bitcoinjs-lib';

export const targetFee = (
  satPerVbyte: number,
  constructTx: (fee: number) => Transaction,
) => {
  const tx = constructTx(1);
  return constructTx(Math.ceil(tx.virtualSize() * satPerVbyte + 1));
};
