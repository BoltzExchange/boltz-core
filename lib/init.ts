import { initEccLib } from 'bitcoinjs-lib';
import { TinySecp256k1Interface } from 'bitcoinjs-lib/src/types';

export const init = (eccLib: TinySecp256k1Interface) => initEccLib(eccLib);
