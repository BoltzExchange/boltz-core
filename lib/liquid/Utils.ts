import { confidential, TxOutput } from 'liquidjs-lib';
import { confidentialLiquid } from './Init';

export const getOutputValue = (
  output: TxOutput & {
    blindingPrivKey?: Buffer;
  },
): number => {
  return output.blindingPrivKey
    ? Number(
        confidentialLiquid.unblindOutputWithKey(output, output.blindingPrivKey)
          .value,
      )
    : confidential.confidentialValueToSatoshi(output.value);
};
