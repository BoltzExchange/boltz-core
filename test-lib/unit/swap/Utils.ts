import Errors from '../../../lib/consts/Errors';
import { OutputType } from '../../../lib/consts/Enums';
import { p2wshOutput, p2shOutput, p2shP2wshOutput } from '../../../lib/swap/Scripts';

export const getScriptHashFunction = (type: OutputType): (scriptHex: Buffer) => Buffer => {
  switch (type) {
    case OutputType.Bech32: return p2wshOutput;
    case OutputType.Legacy: return p2shOutput;
    case OutputType.Compatibility: return p2shP2wshOutput;
    case OutputType.Taproot: throw Errors.TAPROOT_NOT_SUPPORTED;
  }
};
