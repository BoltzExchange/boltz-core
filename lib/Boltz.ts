import Networks from './consts/Networks';
import * as Scripts from './swap/Scripts';
import { OutputType } from './consts/Enums';
import * as SwapUtils from './swap/SwapUtils';
import { swapScript } from './swap/SwapScript';
import { detectSwap } from './swap/SwapDetector';
import { constructClaimTransaction } from './swap/Claim';
import { detectPreimage } from './swap/PreimageDetector';
import { constructRefundTransaction } from './swap/Refund';
import { estimateFee, estimateSize } from './FeeCalculator';
import { ScriptElement, TransactionOutput } from './consts/Types';

export {
  Networks,
  OutputType,
  ScriptElement,
  TransactionOutput,

  Scripts,

  swapScript,

  detectSwap,
  detectPreimage,

  estimateFee,
  estimateSize,

  constructClaimTransaction,
  constructRefundTransaction,

  SwapUtils,
};
