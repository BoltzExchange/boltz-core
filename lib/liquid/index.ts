import { init } from './Init';
import Errors from './consts/Errors';
import Networks from './consts/Networks';
import { getOutputValue } from './Utils';
import { constructClaimTransaction } from './swap/Claim';
import { constructRefundTransaction } from './swap/Refund';

export {
  Errors,
  Networks,
  init,
  getOutputValue,
  constructClaimTransaction,
  constructRefundTransaction,
};
