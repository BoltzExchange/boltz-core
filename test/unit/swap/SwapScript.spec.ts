// tslint:disable:max-line-length
import swapScript from '../../../lib/swap/SwapScript';
import { getHexString, getHexBuffer } from '../../../lib/Utils';

describe('SwapScript', () => {
  const preimageHash = getHexBuffer('53ada8e6de01c26ff43040887ba7b22bddce19f8658fd1ba00716ed79d15cd5e');
  const destinationPublicKey = getHexBuffer('03f8109578aae1e5cfc497e466cf6ae6625497cd31886e87b2f4f54f3f0f46b539');

  const refundPublicKey = getHexBuffer('03ec0c1e45b709d708cd376a6f2daf19ac27be229647780d592e27d7fb7efb207a');

  const timeoutBlockHeight = 515924;

  test('should get a swap script', () => {
    const testData = {
      args: {
        preimageHash,
        destinationPublicKey,
        refundPublicKey,
        timeoutBlockHeight,
      },
      result: 'a914e2ac8cb97af3d59b1c057db4b0c4f9aa12a9127387632103f8109578aae1e5cfc497e466cf6ae6625497cd31886e87b2f4f54f3f0f46b539670354df07b1752103ec0c1e45b709d708cd376a6f2daf19ac27be229647780d592e27d7fb7efb207a68ac',
    };

    const result = swapScript(
      testData.args.preimageHash,
      testData.args.destinationPublicKey,
      testData.args.refundPublicKey,
      testData.args.timeoutBlockHeight,
    );

    expect(getHexString(result)).toEqual(testData.result);
  });
});
