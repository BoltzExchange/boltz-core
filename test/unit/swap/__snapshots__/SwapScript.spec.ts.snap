// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SwapScript should get a swap script 1`] = `"a914e2ac8cb97af3d59b1c057db4b0c4f9aa12a9127387632103f8109578aae1e5cfc497e466cf6ae6625497cd31886e87b2f4f54f3f0f46b539670354df07b1752103ec0c1e45b709d708cd376a6f2daf19ac27be229647780d592e27d7fb7efb207a68ac"`;
